documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/ContentDocs.java

```diff
@@ -0,0 +1,148 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.docs.programming;
+
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.content.AsyncContent;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.FutureCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("unused")
+public class ContentDocs
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ContentDocs.class);
+
+    // tag::echo[]
+    void echo(Content.Source source, Content.Sink sink, Callback callback)
+    {
+        Callback echo = new Callback()
+        {
+            private Content.Chunk chunk;
+
+            public void succeeded()
+            {
+                while (true)
+                {
+                    // release previous chunk
+                    if (chunk != null)
+                    {
+                        chunk.release();
+                        // complete if it was the last
+                        if (chunk.isLast())
+                        {
+                            callback.succeeded();
+                            return;
+                        }
+                    }
+
+                    // read the next chunk
+                    chunk = source.read();
+
+                    if (chunk == null) // if no chunk, demand more
+                        source.demand(this::succeeded);
+                    else if (chunk.hasRemaining()) // if chunk has content, write it to the sink
+                        sink.write(chunk.isLast(), chunk.getByteBuffer(), this);
+                    else if (Content.Chunk.isError(chunk)) // if it is an error fail the callback
+                    {
+                        if (chunk.isLast())
+                            callback.failed(chunk.getError());
+                        else
+                        {
+                            LOG.warn("transient error", chunk.getError());
+                            continue;
+                        }
+                    }
+                    else // otherwise continue with another chunk
+                        continue;
+                    return;
```

Comment:
Very hard to follow.

I would prefer `source.read()` to be the first statement inside the loop, then check for null and return, then check for error and handle+return, then it is a normal chunk and we can write it.

The code above is also weird in case of a last empty chunk, where the write is skipped -- it is really hard to "see" that succeeding the callback also does the last write; for example, the loop above would probably not work for a proxy.

Reply:
@sbordet How about now?